thread/thread.cpp

```diff
@@ -1749,30 +1749,27 @@ R"(
     {
         if (count == 0) return 0;
         splock.lock();
+        DEFER(splock.unlock());
         CURRENT->semaphore_count = count;
         int ret = 0;
         while (!try_subtract(count)) {
             ret = waitq::wait_defer(timeout, spinlock_unlock, &splock);
-            ERRNO err;
-            splock.lock();
-            if (ret < 0) {
+            splock.lock();  // assuming errno NOT changed
+            if (unlikely(ret < 0)) {    // got interrupted
                 CURRENT->semaphore_count = 0;
                 // when timeout, we need to try to resume next thread(s) in q
+                ERRNO err;
                 if (err.no == ETIMEDOUT) try_resume();
```

Comment:
Once a wait timeout expired, it will always try resume no matter the `in_order_resume` is.

Is that works as expected?